include/envoy/http/async_client.h

```diff
@@ -38,7 +38,40 @@ class AsyncClient {
   };
 
   /**
-   * An in-flight HTTP request
+   * Notifies caller of async HTTP stream status.
+   */
+  class StreamCallbacks {
+  public:
+    virtual ~StreamCallbacks() {}
+
+    /**
+     * Called when the async HTTP stream receives headers.
+     * @param headers the headers received
+     * @param end_stream whether the response is header only
+     */
+    virtual void onHeaders(HeaderMapPtr&& headers, bool end_stream) PURE;
+
+    /**
+     * Called when the async HTTP stream receives data.
+     * @param data the data received
+     * @param end_stream whether the data is the last data frame
+     */
+    virtual void onData(Buffer::Instance& data, bool end_stream) PURE;
```

Comment:
I think I mostly agree about StreamDecoderFilter but there are many aspects of StreamDecoderFilterCallbacks that people are unlikely to need or want. 

The fact that an explicit 'reset' handler must be installed

https://github.com/lyft/envoy/blob/89de442d82a1727aa7f53604748e953cc27b4849/include/envoy/http/filter.h#L69

does make the API less obvious than this one.